### ISSUE TITLE
Failed to  test the tutorial in the clipper  source code version 

### ISSUE THREAD

**Issue by frank2wang87** [ucbrise/clipper#117]
I could run  the tutorial successfully in the clipper docker version. But it always reports "No containers found for model tf_cifar:1" when I run the clipper source code version.

The source code clipper was generated in Ubuntu 14.04. (** can clippper work on Ubuntu 14.04?**)

My working steps  of the clipper source code version are the same as the clipper docker version ,which are  marked as following:
1. run  ./bin/start_clipper.sh  to start clipper.
2. Run the following python code to deploy new model container

################################################################3
cifar_loc="./data"
import sys
import os
sys.path.append(os.path.abspath('../../management/'))
import clipper_manager as cm
import cifar_utils
test_x, test_y = cifar_utils.filter_data(*cifar_utils.load_cifar(cifar_loc, cifar_filename="cifar_test.data", norm=True))

user = ""
key = ""
host = "localhost"
clipper = cm.Clipper(host, user, key)

_**## The clipper has been stared in source code , so not to start the docker version here**_
##clipper.start()

app_name = "cifar_demo"
candidate_models = [
    {"model_name": "tf_cifar", "model_version": 1},
]

clipper.register_application(
    app_name,
    candidate_models,
    "doubles",
    "EXP4",
    slo_micros=20000)

model_added = clipper.deploy_model(
    "tf_cifar",
    1,
    os.path.abspath("tf_cifar_model"),
    "clipper/tf_cifar_container:latest",
    ["cifar", "tf"],
    "doubles",
    num_containers=1
)
print("Model deploy successful? {success}".format(success=model_added))
###############################################################
After executing these codes, it outputs:
Found clipper/tf_cifar_container:latest in Docker hub
Copied model data to host
Published model to Clipper
Model deploy successful? True

So the model container was deployed to clipper successfully.
##################################################################

However, the outputs for "  ./bin/start_clipper.sh" are:
[22:38:54.493][info]       [REDIS] Successfully issued command "SELECT 2"
[22:38:54.493][info]       [REDIS] Successfully issued command "HMSET tf_cifar:1 model_name tf_cifar model_version 1 load 0.000000 input_type doubles labels cifar,tf container_name clipper/tf_cifar_container:latest model_data_path /tmp/clipper-models/tf_cifar/1/tf_cifar_model"
[22:39:24.963][error]      [REDIS] Error with command "GET cifar_demo:0:0": 
[22:39:24.964][info]  [QUERYPR...] Found 1 tasks
_**[22:39:24.964][info]  [TASKEXE...] No active containers found for model tf_cifar:1**_

AND if I sent a prediction request, it was the random result. Obviously, the new model container wasn't used. 

In addition,  the following RPC related information was found when I deployed the new container in the clipper docker version. But it can't be found in the clipper source code version.  So I guess it maybe some problems to the RPC interface. What's the exact problem ?
**_[14:33:04.732][info]         [RPC] Found message to receive
[14:33:04.732][info]         [RPC] New container connected
[14:33:04.732][info]         [RPC] Container added_**
[14:33:04.733][info]       [REDIS] Successfully issued command "SELECT 3"
[14:33:04.733][info]       [REDIS] Successfully issued command "HMSET tf_cifar:1:0 model_id tf_cifar:1 model_name tf_cifar model_version 1 model_replica_id 0 zmq_connection_id 0 batch_size 1 input_type doubles"

**Comment by dcrankshaw** [comment]
I believe this is happening because you haven't downloaded the TensorFlow model. Are there any files in the `examples/tutorial/tf_cifar_model` directory?

The model deployment itself was successful because the Clipper manager library doesn't know that you were missing files and it was able to start the container fine. But when the container started, it was missing the model files it was expecting so it immediately exited.

Here are the instructions from the tutorial (`tutorial/tutorial_part_two.ipynb`) on how to download the model:
> There is a pre-trained TensorFlow model stored in the repo using git-lfs. Once you install git-lfs, you can download the model with the command `git lfs pull`.

Let me know if that solves the problem.

**Comment by frank2wang87** [comment]
@ dcrankshaw
hi, dcrankshaw.  I checked the `examples/tutorial/tf_cifar_model` directory and I could find the tensorflow model. In fact, the tensorflow model has been successfully copied to `/tmp/clipper-models/tf_cifar/1/ tf_cifar_model` directory.

When I modified the CLIPPER_IP variable to the explicit IP address(`192.168.0.3`) in the add_container function of clipper_manager.py, this problem was solved. Why?


add_container_cmd = (
                "docker run -d --network={nw} -v {path}:/model:ro "
                "-e \"CLIPPER_MODEL_NAME={mn}\" -e \"CLIPPER_MODEL_VERSION={mv}\" "
                **_#"-e \"CLIPPER_IP=query_frontend\" -e \"CLIPPER_INPUT_TYPE={mip}\" "_**
               **_"-e \"CLIPPER_IP=192.168.0.3\" -e \"CLIPPER_INPUT_TYPE={mip}\" "_**
                "{image}".format(
                    path=model_data_path,
                    nw=DOCKER_NW,
                    image=image_name,
                    mn=model_name,
                    mv=model_version,
                    mip=model_input_type))

Unfortunately, I came across another problem both in clipper docker version and source code version. When I deployed a new container which predicted more than two classes(e.g., 1000 classes for imagenet dataset, the prediction result is 0-999)，the container could predict the right class(i.e., class number 100), while the clipper always returned class number 1.
 In other words, the input data path(`REST prediction request->clipper->RPC->model container`)worked well,but there was some problem in the output data path(`model container-> RPC->clipper -> REST response`).

I read #118  and set an long latency requirements(e.g, 150ms), but it remained the same problem.

`

**Comment by dcrankshaw** [comment]
The current selection policies are kind of broken and only support binary classification right now. We are removing them in #89 which will fix this problem.

**Comment by frank2wang87** [comment]
@dcrankshaw. Does the clipper current support any other AI application(e.g., object detection, semantic segmentation) besides binary image classification?

**Comment by dcrankshaw** [comment]
Conceptually, Clipper supports any type of AI application. In practice, models are currently constrained to returning a single float per prediction. We are in the process of relaxing this constraint to allow models to return arbitrary strings (with the recommendation that they return JSON). This feature is tracked in [CLIPPER-143](https://clipper.atlassian.net/browse/CLIPPER-143).

**Comment by frank2wang87** [comment]
@dcrankshaw. Sounds Great! When will  the new version clipper be published to support to return arbitrary strings?

**Comment by dcrankshaw** [comment]
I'm not sure yet. @Corey-Zumar is going to spend some time on it this week, so hopefully early next week. For sure in the next week or two.

**Comment by chakpongchung** [comment]
@dcrankshaw 

" In practice, models are currently constrained to returning a single float per prediction."

If I want to add support for object detection with tensorflow, then one single float return is not enough. Ideally it can return a json file. Where is this constraint from? 

For sending images as input to the object detection, can Clipper support images uploading? So far I am thinking using gRPC to upload the image by chunking, not sure whether this is a recommended approach. The default message size for protobuf is 64kb only but the image is from 1MB to 6MB. Or I should use other ways to send in the image to be detected?

 I would like to add this support with some guidance as we have this business need. In the end we want to have the classification and detection module ready to provide microservices for downstream application. We have our own trained model in tensorflow and caffe format.

**Comment by dcrankshaw** [comment]
Hi @chakpongchung. It looks like that documentation is out of date. Where are you quoting from?

In fact, models must now return predictions as strings. These are commonly JSON strings, but aren't constrained to be valid JSON. The hope is that this gives users a lot of flexibility in terms of what their models product. So this should already meet your needs.

As far as sending the images to Clipper, currently we support predictions via a REST interface, which supports both integer/float vectors (if you already have your images as pixels), or base64 encoded byte strings (if your images are JPEGs or some other binary format). We don't have a GRPC frontend right now, although we recently implemented a prototype GRPC frontend for some performance evaluation (if you're curious, [here's the code](https://github.com/dcrankshaw/clipper/blob/78f6783fa3233ca6802f7eab35935f4f5a9f0b0a/src/rpc_frontend/src/grpc_frontend.hpp)). We actually didn't find GRPC to be all that much faster than the REST frontend in our benchmarks, which included uploading ~1MB images for prediction.

I would recommend starting by base64 encoding your messages and uploading them using the REST interface. If that ends up being too slow for your purposes, I'd be happy to guide you through implementing a GRPC frontend based on the prototype linked above. We actually already have #238 to track this, although with no immediate plans to work on it.

Let me know if you have more questions.

**Comment by chakpongchung** [comment]
@dcrankshaw Thanks for the feedback!

From the paper and here it seems you are using base64  or int/float vectors for file transfer first, waiting for batch construction and then send the query using RPC? The images are pre-uploaded first  before the  batch size is reached or uploaded as arguments along with the RPC call after the batch size is reached ? 

For detection, my model is trained using tensorflow. It seems detection template is not implemented in Clipper yet. If I need to implement it for business need, could you point out where should get started?

A brief performance summary, on 1080TI, my detection code runs at most 15fps since it is cascading two detection models A and B together. If the things are not detected from A, there is no need to use the result from A as input to B.  The result from detection model A will also go as input to classification model C, which runs faster than 30fps. The current goal is to prioritize the speed of detection A since the pipeline is expected to run for a video stream.

A->B
A ->C

What would be the right strategy to implement that in Clipper?
